engines/sci/sound/decoders/sol.cpp

```diff
@@ -0,0 +1,269 @@
+	// Reading an odd number of 8-bit samples will result in a loss of samples
+	// since one byte represents two samples and we do not store the second
+	// nibble in this case; it should never happen in reality
+	assert(S16BIT || (numSamples % 2) == 0);
+
+	const int samplesPerByte = S16BIT ? 1 : 2;
+
+	int bytesToRead = numSamples / samplesPerByte;
+	if (_stream->pos() + bytesToRead > _rawDataSize) {
+		bytesToRead = _rawDataSize - _stream->pos();
+	}
+
+	if (S16BIT) {
+		_dpcm16Carry = deDPCM16(buffer, *_stream, bytesToRead, _dpcm16Carry);
+	} else {
+		_dpcm8Carry = deDPCM8(buffer, *_stream, bytesToRead, _dpcm8Carry);
+	}
+
+	const int samplesRead = bytesToRead * samplesPerByte;
+	return samplesRead;
+}
+
+template <bool STEREO, bool S16BIT>
+bool SOLStream<STEREO, S16BIT>::isStereo() const override {
+	return STEREO;
+}
+
+template <bool STEREO, bool S16BIT>
+int SOLStream<STEREO, S16BIT>::getRate() const override {
+	return _sampleRate;
+}
+
+template <bool STEREO, bool S16BIT>
+bool SOLStream<STEREO, S16BIT>::endOfData() const override {
+	return _stream->eos() || _stream->pos() >= _dataOffset + _rawDataSize;
+}
+
+template <bool STEREO, bool S16BIT>
+bool SOLStream<STEREO, S16BIT>::rewind() override {
+	_dpcm8Carry = 0x80;
+	_dpcm16Carry = 0;
+	return seek(0);
+}
+
+Audio::SeekableAudioStream *makeSOLStream(Common::SeekableReadStream *stream, DisposeAfterUse::Flag disposeAfterUse) {
+
+	// TODO: Might not be necessary? Makes seeking work, but
+	// not sure if audio is ever actually seeked in SSCI.
+	const int32 initialPosition = stream->pos();
+
+	byte header[6];
+	if (stream->read(header, sizeof(header)) != sizeof(header)) {
+		return nullptr;
+	}
+
+	if (header[0] != 0x8d || READ_BE_UINT32(header + 2) != MKTAG('S', 'O', 'L', 0)) {
+		return nullptr;
+	}
+
+	const uint8 headerSize = header[1];
+	const uint16 sampleRate = stream->readUint16LE();
+	const byte flags = stream->readByte();
+	const uint32 dataSize = stream->readUint32LE();
+
+	if (flags & kCompressed) {
+		if (flags & kStereo && flags & k16Bit) {
+			return new SOLStream<true, true>(new Common::SeekableSubReadStream(stream, initialPosition, initialPosition + dataSize, disposeAfterUse), disposeAfterUse, headerSize, sampleRate, dataSize);
+		} else if (flags & kStereo) {
+			return new SOLStream<true, false>(new Common::SeekableSubReadStream(stream, initialPosition, initialPosition + dataSize, disposeAfterUse), disposeAfterUse, headerSize, sampleRate, dataSize);
+		} else if (flags & k16Bit) {
+			return new SOLStream<false, true>(new Common::SeekableSubReadStream(stream, initialPosition, initialPosition + dataSize, disposeAfterUse), disposeAfterUse, headerSize, sampleRate, dataSize);
+		} else {
+			return new SOLStream<false, false>(new Common::SeekableSubReadStream(stream, initialPosition, initialPosition + dataSize, disposeAfterUse), disposeAfterUse, headerSize, sampleRate, dataSize);
+		}
```

Comment:
Because using a template allows elimination of conditionals at runtime. It’s not hot enough code to likely matter, but there’s no real ergonomic benefit to not using a template here.